File: backend/aura/workflow/governor.py
"""NodeGovernor — port of governor.ts: the engine's ONE door to the Fabric."""
from __future__ import annotations

import re
from typing import Any, Callable

from .nodes_core import bindings

MAX_OUTPUT_TEXT = 60_000
_SECRET_REF = re.compile(r"\{\{\s*secret:([\w.-]+)\s*\}\}")


def text_of(result: dict) -> str:
    out = result.get("output")
    if isinstance(out, str):
        return out
    if isinstance(out, dict):
        if isinstance(out.get("stdout"), str):
            return out["stdout"]
        if isinstance(out.get("text"), str):
            return out["text"]
    return result.get("detail") or ""


def evidence_of(result: dict, node_id: str | None = None) -> dict:
    ev = {
        "invocationId": result["invocationId"],
        "capabilityId": result["capabilityId"],
        "outcome": result["outcome"],
        "decision": result["policy"]["decision"],
        "decisionRule": result["policy"]["rule"],
        "risk": result["policy"]["risk"],
        "verified": result["verification"]["passed"],
        "at": result["endedAt"],
        "durationMs": result["durationMs"],
    }
    if result.get("approvalId"):
        ev["approvalId"] = result["approvalId"]
    if node_id:
        ev["nodeId"] = node_id
    return ev


class NodeGovernor:
    def __init__(self, fabric, secrets, *, workflow_id, run_id, project_id,
                 project_path, actor, approved_capabilities=None, timeout_ms=None) -> None:
        self.fabric = fabric
        self.workflow_id = workflow_id
        self.run_id = run_id
        self.project_id = project_id
        self.project_path = project_path
        self.actor = actor
        self.approved_capabilities = approved_capabilities
        self.timeout_ms = timeout_ms
        self._bindings = bindings()
        known = list(getattr(secrets, "known_values", lambda: [])())
        self._known = known

    def redact(self, text: str) -> str:
        for v in self._known:
            if v:
                text = text.replace(v, "<redacted>")
        return text

    async def run(self, node: dict, ctx: dict, input: dict,
                  interpolate: Callable[[str], str]) -> dict:
        planner = self._bindings.get(node["type"])
        if not planner:
            return {"kind": "failed", "text": "", "summary": "not governed",
                    "error": f'"{node["type"]}" is classified as a governed node but has no capability binding, so nothing was run.'}
        try:
            plan = planner(ctx, input, dict(node.get("config") or {}), interpolate)
        except Exception as e:  # noqa: BLE001
            return {"kind": "failed", "text": "", "summary": "misconfigured",
                    "error": self.redact(str(e))}
        if plan is None:
            return {"kind": "no-effect", "text": input.get("text") or "", "summary": "nothing to do"}

        resolved_input: dict[str, Any] = {}
        try:
            for k, v in (plan.get("input") or {}).items():
                if isinstance(v, str):
                    resolved_input[k] = self._resolve_secret(v)
                elif isinstance(v, dict):
                    resolved_input[k] = {hk: self._resolve_secret(hv) if isinstance(hv, str) else hv
                                         for hk, hv in v.items()}
                else:
                    resolved_input[k] = v
        except Exception as e:  # noqa: BLE001
            return {"kind": "failed", "text": "", "summary": "missing secret",
                    "error": self.redact(str(e))}

        result = await self.fabric.invoke(plan["capabilityId"], resolved_input, {
            "actor": self.actor,
            "projectId": self.project_id,
            "cwd": self.project_path,
            "workflowId": self.workflow_id,
            "runId": self.run_id,
            "workflowNodeId": node["id"],
            **({"approvedCapabilities": self.approved_capabilities}
               if self.approved_capabilities else {}),
            **({"timeoutMs": self.timeout_ms} if self.timeout_ms else {}),
        })

        performed_by = result.get("output", {}).get("nodeId") if isinstance(result.get("output"), dict) else None
        evidence = evidence_of(result, performed_by if isinstance(performed_by, str) else None)
        raw = text_of(result)
        text = self.redact(raw if len(raw) <= MAX_OUTPUT_TEXT else raw[:MAX_OUTPUT_TEXT] + "\n…(truncated)")

        outcome = result["outcome"]
        if outcome == "succeeded":
            shaped = plan.get("shape")(text) if plan.get("shape") else None
            return {"kind": "succeeded",
                    "text": shaped["text"] if shaped and "text" in shaped else text,
                    "data": shaped.get("data") if shaped and "data" in shaped else result.get("output"),
                    "files": shaped.get("files") if shaped else None,
                    "summary": self.redact(shaped["summary"] if shaped and shaped.get("summary") else plan["describe"]),
                    "evidence": evidence}
        if outcome == "unverified":
            return {"kind": "failed", "text": text, "summary": "unverified",
                    "error": self.redact(result.get("detail") or ""), "evidence": evidence}
        if outcome == "denied":
            return {"kind": "denied", "text": "",
                    "summary": f"denied · {result['policy']['rule']}",
                    "error": self.redact(result.get("detail") or ""), "evidence": evidence}
        if outcome == "awaiting-approval":
            approval = None
            if result.get("approvalId"):
                approval = {"requestId": result["approvalId"],
                            "capabilityId": plan["capabilityId"],
                            "requestedAt": result["startedAt"],
                            "summary": self.redact(plan["describe"])}
            return {"kind": "awaiting-approval", "text": "", "summary": "waiting on you",
                    "evidence": evidence, "approval": approval}
        if outcome == "unsupported":
            return {"kind": "failed", "text": "", "summary": "not supported",
                    "error": self.redact(result.get("detail") or ""), "evidence": evidence}
        return {"kind": "failed", "text": text, "summary": "failed",
                "error": self.redact(result.get("detail") or ""), "evidence": evidence}

    def _resolve_secret(self, template: str) -> str:
        """Resolve {{secret:NAME}} via the injected secrets store."""
        store_resolve = getattr(self, "_resolve_one", None)

        def repl(m):
            if store_resolve:
                return store_resolve(m.group(1))
            raise RuntimeError(f'secret "{m.group(1)}" is not configured')

        return _SECRET_REF.sub(repl, template)


def create_governor(**deps) -> NodeGovernor:
    return NodeGovernor(**deps)

File: packages/ai-service/src/workflow/governor.ts
/**
 * governor — the workflow engine's one door to the Capability Fabric.
 * ==================================================================
 * The engine in `engine.ts` knows nothing about capabilities, policy,
 * approval or audit. It knows that some nodes are governed, and it hands
 * those to a `NodeGovernor`. This file is the only implementation, and it
 * does four things and no others:
 *
 *   1. plan  — node type + config → capability id + arguments (`governed.ts`)
 *   2. inject— resolve `{{secret:NAME}}` into the arguments, last moment
 *   3. invoke— `fabric.invoke(...)`, with the run's correlation keys
 *   4. map   — one `InvocationResult` → one `GovernedOutcome`
 *
 * It makes no decision the Fabric could have made. Notice in particular
 * what is NOT here: no risk check, no allow-list, no "is this safe",
 * no retry (the Fabric's bounded recovery already owns that), no audit
 * write. Every one of those would be a second authority.
 *
 * The layering is deliberate: keeping the Fabric behind an interface means
 * the engine can be exercised with a stub governor in tests that must NOT
 * touch a real machine, while the production path has no bypass at all —
 * a governed node with no governor attached refuses to run rather than
 * falling back to a direct effect.
 */

import type { CapabilityFabric, InvocationResult } from '@aura/capability-fabric';
import { BINDINGS, isGoverned } from './governed';
import type { RunCtx } from './nodes';
import type { EvidenceRef } from './run/types';
import type { NodeIO, WfNode } from './types';
import type { SecretStore } from '../secrets';

export interface GovernedOutcome {
  kind: 'succeeded' | 'failed' | 'denied' | 'awaiting-approval' | 'no-effect';
  /** Text to pass downstream. Already redacted. */
  text: string;
  data?: unknown;
  files?: string[];
  /** Short node-face summary. Already redacted. */
  summary: string;
  /** Set on `failed` / `denied`. Already redacted. */
  error?: string;
  /** The audit-trail pointer, when an invocation actually happened. */
  evidence?: EvidenceRef;
  /** Set on `awaiting-approval`. */
  approval?: { requestId: string; capabilityId: string; requestedAt: string; summary: string };
}

export interface NodeGovernor {
  /** Run one governed node through the Fabric. */
  run(node: WfNode, ctx: RunCtx, input: NodeIO, interpolate: (t: string) => string): Promise<GovernedOutcome>;
  /** Scrub known secret values out of anything about to be recorded. */
  redact(text: string): string;
}

export interface GovernorDeps {
  fabric: CapabilityFabric;
  secrets: SecretStore;
  workflowId: string;
  runId: string;
  projectId: string;
  projectPath: string;
  /** Who the audit trail should hold responsible. */
  actor: { kind: 'human' | 'agent' | 'system'; id: string };
  /**
   * Capability ids the human authorized for THIS run, if any.
   *
   * Passed straight to the Fabric's per-invocation grant. It is not a
   * standing permission and it is not consulted here — the Fabric still
   * evaluates policy, still applies every hard floor, and still spends the
   * grant once. A run authorized for `filesystem.write` cannot use that to
   * reach `terminal.execute`.
   */
  approvedCapabilities?: string[];
  /** Per-node wall clock handed to the executor. */
  timeoutMs?: number;
}

/** Text kept from an executor's output. Bounded so a run record cannot bloat. */
const MAX_OUTPUT_TEXT = 60_000;

/** Pull the human-usable text out of an executor payload without guessing. */
function textOf(result: InvocationResult): string {
  const out = result.output;
  if (typeof out === 'string') return out;
  if (out && typeof out === 'object') {
    const o = out as { stdout?: unknown; text?: unknown };
    if (typeof o.stdout === 'string') return o.stdout;
    if (typeof o.text === 'string') return o.text;
  }
  return result.detail;
}

function evidenceOf(result: InvocationResult, nodeId?: string): EvidenceRef {
  return {
    invocationId: result.invocationId,
    capabilityId: result.capabilityId,
    outcome: result.outcome,
    decision: result.policy.decision,
    decisionRule: result.policy.rule,
    risk: result.policy.risk,
    verified: result.verification.passed,
    approvalId: result.approvalId,
    nodeId,
    at: result.endedAt,
    durationMs: result.durationMs,
  };
}

export function createGovernor(deps: GovernorDeps): NodeGovernor {
  // Built once per run: the values are decrypted a single time and live
  // only as long as the run does.
  const redact = deps.secrets.redactor();

  return {
    redact,

    async run(node, ctx, input, interpolate) {
      const planner = BINDINGS[node.type];
      if (!isGoverned(node.type) || !planner) {
        // Reaching here means the engine's classification and this file's
        // bindings disagree. That is a build error, and failing closed is
        // the only safe reading of it — the alternative is performing an
        // ungoverned effect because a table was incomplete.
        return { kind: 'failed', text: '', summary: 'not governed', error: `"${node.type}" is classified as a governed node but has no capability binding, so nothing was run.` };
      }

      let plan;
      try {
        plan = planner(ctx, input, { ...node.config }, interpolate);
      } catch (error) {
        const message = redact((error as Error).message);
        return { kind: 'failed', text: '', summary: 'misconfigured', error: message };
      }
      if (!plan) return { kind: 'no-effect', text: input.text, summary: 'nothing to do' };

      /* Secret injection — the last moment before the Fabric, and the only
         place a value exists in plaintext. `plan.describe` is built from
         the UNRESOLVED text, so the approval prompt and the run log show
         the reference, never the credential. */
      let resolvedInput: Record<string, unknown>;
      try {
        resolvedInput = Object.fromEntries(
          Object.entries(plan.input).map(([k, v]) => {
            if (typeof v === 'string') return [k, deps.secrets.resolve(v).text];
            if (v && typeof v === 'object' && !Array.isArray(v)) {
              return [k, Object.fromEntries(
                Object.entries(v as Record<string, unknown>).map(([hk, hv]) =>
                  [hk, typeof hv === 'string' ? deps.secrets.resolve(hv).text : hv]),
              )];
            }
            return [k, v];
          }),
        );
      } catch (error) {
        return { kind: 'failed', text: '', summary: 'missing secret', error: redact((error as Error).message) };
      }

      const result = await deps.fabric.invoke(plan.capabilityId, resolvedInput, {
        actor: deps.actor,
        projectId: deps.projectId,
        cwd: deps.projectPath,
        workflowId: deps.workflowId,
        runId: deps.runId,
        workflowNodeId: node.id,
        approvedCapabilities: deps.approvedCapabilities,
        timeoutMs: deps.timeoutMs,
      });

      const performedBy = (result.output as { nodeId?: unknown } | undefined)?.nodeId;
      const evidence = evidenceOf(result, typeof performedBy === 'string' ? performedBy : undefined);
      const raw = textOf(result);
      const text = redact(raw.length > MAX_OUTPUT_TEXT ? `${raw.slice(0, MAX_OUTPUT_TEXT)}\n…(truncated)` : raw);

      switch (result.outcome) {
        case 'succeeded': {
          // Shaping runs on already-redacted text, so a shape function can
          // never re-expose a value the redactor removed.
          const shaped = plan.shape ? plan.shape(text) : null;
          return {
            kind: 'succeeded',
            text: shaped?.text ?? text,
            data: shaped && 'data' in shaped ? shaped.data : result.output,
            files: shaped?.files,
            summary: redact(shaped?.summary ?? plan.describe),
            evidence,
          };
        }

        // Ran, but could not be shown to have done what it claimed. Treated
        // as a failure for the same reason the mission path does: an
        // unverified effect the operator believes succeeded is worse than
        // one they know to re-check.
        case 'unverified':
          return { kind: 'failed', text, summary: 'unverified', error: redact(result.detail), evidence };

        case 'denied':
          return { kind: 'denied', text: '', summary: `denied · ${result.policy.rule}`, error: redact(result.detail), evidence };

        case 'awaiting-approval':
          return {
            kind: 'awaiting-approval',
            text: '',
            summary: 'waiting on you',
            evidence,
            approval: result.approvalId
              ? { requestId: result.approvalId, capabilityId: plan.capabilityId, requestedAt: result.startedAt, summary: redact(plan.describe) }
              : undefined,
          };

        case 'unsupported':
          return { kind: 'failed', text: '', summary: 'not supported', error: redact(result.detail), evidence };

        case 'failed':
        default:
          return { kind: 'failed', text, summary: 'failed', error: redact(result.detail), evidence };
      }
    },
  };
}
